pkg/operator/configobservation/apiserver/observe_apiserver.go

```diff
@@ -0,0 +1,299 @@
+	resourceSyncRules := syncActionRules{}
+	var observedNamedCertificates []interface{}
+	for index, namedCertificate := range namedCertificates {
+		observedNamedCertificate := map[string]interface{}{}
+		if len(namedCertificate.Names) > 0 {
+			if err := unstructured.SetNestedStringSlice(observedNamedCertificate, namedCertificate.Names, "names"); err != nil {
+				return previouslyObservedConfig, nil, append(errs, err)
+			}
+		}
+		sourceSecretName := namedCertificate.ServingCertificate.Name
+		if len(sourceSecretName) == 0 {
+			err := fmt.Errorf("spec.servingCerts.namedCertificates[%v].servingCertificate.name cannot be empty", index)
+			recorder.Warningf("ObserveNamedCertificatesFailed", err.Error())
+			return previouslyObservedConfig, nil, append(errs, err)
+		}
+		// pick one of the available target resource names
+		targetSecretName := fmt.Sprintf(namedUserServingCertResourceNameFormat, index)
+
+		// add a sync rule to copy the user specified secret to the operator namespace
+		resourceSyncRules[targetSecretName] = sourceSecretName
+
+		// add the named certificate to the observed config
+		certFile := fmt.Sprintf("/etc/kubernetes/static-pod-resources/secrets/%v/tls.crt", targetSecretName)
+		if err := unstructured.SetNestedField(observedNamedCertificate, certFile, "certFile"); err != nil {
+			return previouslyObservedConfig, nil, append(errs, err)
+		}
+
+		// add key
+		keyFile := fmt.Sprintf("/etc/kubernetes/static-pod-resources/secrets/%v/tls.key", targetSecretName)
+		if err := unstructured.SetNestedField(observedNamedCertificate, keyFile, "keyFile"); err != nil {
+			return previouslyObservedConfig, nil, append(errs, err)
+		}
+
+		observedNamedCertificates = append(observedNamedCertificates, observedNamedCertificate)
+	}
+	if len(namedCertificates) > 0 {
+		if err := unstructured.SetNestedField(observedConfig, observedNamedCertificates, namedCertificatesPath...); err != nil {
+			return previouslyObservedConfig, nil, append(errs, err)
+		}
+	}
+
+	return observedConfig, resourceSyncRules, errs
+}
+
+type apiServerObserver struct {
+	observerFunc  observeAPIServerConfigFunc
+	configPaths   [][]string
+	resourceNames []string
+	resourceType  interface{}
+}
+
+func (o *apiServerObserver) observe(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	var errs []error
+
+	// pick the correct resource sync function
+	resourceSync := listers.ResourceSyncer().SyncSecret
+	if _, ok := o.resourceType.(corev1.ConfigMap); ok {
+		resourceSync = listers.ResourceSyncer().SyncConfigMap
+	}
+
+	// extract the previously observed config. capture, but don't react to, errors.
+	previouslyObservedConfig, errs := extractPreviouslyObservedConfig(existingConfig, o.configPaths...)
+
+	// get the apiserver config
+	apiServer, err := listers.APIServerLister.Get("cluster")
```

Comment:
this does not need a comment, it doesn't say anything else than the one line of code.